### ISSUE TITLE
[CRITICAL] Architectural flaw: Models shouldn't have Train() methods called by optimizers — library-wide impact

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1123]
## 🚨 CRITICAL ARCHITECTURAL FLAW — Library-Wide Impact

This bug affects **every optimizer × model combination** in AiDotNet and makes training unusably slow or functionally broken. A simple regression on 2000 rows takes 5+ minutes when it should take under 1 second.

## Core Problem

AiDotNet's current architecture has optimizers call the model's `Train()` method inside their epoch loop. This is **backwards** compared to how modern ML libraries (PyTorch, TensorFlow, JAX) work.

### What Currently Happens

```
AdamOptimizer.Optimize()                           // epoch loop starts
  └─ for epoch in range(MaxIterations):            // e.g., 1000 iterations
      ├─ CalculateGradient() + UpdateSolution()    // OK — Adam parameter update
      └─ EvaluateSolution(currentSolution)
          └─ PrepareAndEvaluateSolution()
              └─ TrainAndEvaluateSolution()
                  └─ model.Train(x, y)             // ❌ Model trains ITSELF from scratch
                      ├─ MultipleRegression: closed-form solve — same result every time
                      └─ NeuralNetworkBase: TrainWithTape() — full internal backprop loop
```

**Result:** The external optimizer's epoch loop triggers `model.Train()` every epoch, which does a COMPLETE from-scratch training internally. So:
- Closed-form models: the same linear system gets re-solved 1000 times with identical results
- Neural networks: the NN trains itself internally while the external Adam optimizer ALSO tries to update parameters → the two fight each other and the external Adam's work is wasted

## How PyTorch Does It (Correctly)

```python
model = nn.Linear(4, 1)                    # Model is just a FUNCTION
optimizer = torch.optim.Adam(model.parameters(), lr=0.01)

for epoch in range(num_epochs):
    predictions = model(x)                 # Forward pass
    loss = loss_fn(predictions, y)
    loss.backward()                        # Gradients computed by autograd
    optimizer.step()                       # Parameter update
    optimizer.zero_grad()
```

Key principles:
1. **Models are functions** — they have `forward()` for predictions. No `train()` method that does optimization.
2. **Optimizers only update parameters** — they don't know how to train models.
3. **The user writes the loop** (or uses a training utility that does).
4. **Closed-form models** (like linear regression) don't use optimizers at all — you call `torch.linalg.lstsq(x, y)` directly.

### Week 1 AI Training Course (Production Use Case)
- Homework: train regression on 2000 building records, 4 features
- Expected: <1 second (closed-form solve)
- Actual: 5+ minutes, appears to hang, never completes with default `MaxIterations=1000`
- Students report the program 'runs but does nothing' — GPU diagnostics appear then it silently stalls

### The Internal `Train()` Method Implementations

**`MultipleRegression.Train()` (line 109 of MultipleRegression.cs):**
```csharp
public override void Train(Matrix<T> x, Vector<T> y)
{
    if (Options.UseIntercept) x = x.AddConstantColumn(NumOps.One);
    var xTx = x.Transpose().Multiply(x);
    var regularizedXTx = xTx.Add(Regularization.Regularize(xTx));
    var xTy = x.Transpose().Multiply(y);
    var solution = SolveSystem(regularizedXTx, xTy);
    // ... assigns solution to Coefficients/Intercept
}
```
This is a **closed-form solve** — one shot, no iteration needed. Called 1000 times by optimizer = 1000 redundant matrix inversions.

**`NeuralNetworkBase.Train()` (line 2374):**
```csharp
public virtual void Train(Tensor<T> input, Tensor<T> expectedOutput)
{
    SetTrainingMode(true);
    var trainableParams = Training.TapeTrainingStep<T>.CollectParameters(Layers);
    if (trainableParams.Count > 0)
    {
        TrainWithTape(input, expectedOutput, optimizer: null);  // Internal training loop
    }
    // ...
}
```
The NN trains itself internally. So when the external `AdamOptimizer` calls this, the NN runs its own backprop. The external Adam's gradient updates are then OVERWRITTEN by the NN's internal training.

### Existing Early Stopping Doesn't Help

`OptimizationAlgorithmOptions.UseEarlyStopping = true` is the default, but:
- `EarlyStoppingPatience = 10` means minimum 10 epochs before it can even trigger
- 10 epochs × (expensive model.Train() each) = already way too slow
- For closed-form models, the fitness scores are identical every iteration, so 'no improvement' logic keeps the history but doesn't trigger fast enough

## Affected Components

**Every model type used through the `AiModelBuilder` facade with an optimizer:**
- ❌ `MultipleRegression` — closed-form, wasted iteration
- ❌ `RidgeRegression`, `LassoRegression`, `ElasticNet` — same closed-form issue
- ❌ `Transformer` — NN trains itself, external optimizer fights it
- ❌ All `NeuralNetworkBase` subclasses (CNN, RNN, LSTM, Autoencoder, etc.)
- ❌ Any model with internal `Train()` logic

### Option 1: Separate closed-form from iterative (cleanest)

1. Add `IClosedFormModel` marker interface for models with analytical solutions
2. `AiModelBuilder.BuildAsync()` detects this interface:
   - If closed-form: call `model.Solve(x, y)` directly, skip optimizer entirely
   - If iterative: use optimizer normally

### Option 2: Redesign optimizer/model contract (PyTorch-style)

1. Models expose only `Forward(x)` and `Backward(grad)` — no `Train()`
2. Optimizers do the training loop:
   - `predictions = model.Forward(x)`
   - `gradient = lossFn.Derivative(predictions, y)`
   - `paramGrads = model.Backward(gradient)`
   - `model.SetParameters(optimizer.Step(paramGrads))`
3. No internal training inside models

### Option 3: Quick Fix (temporary)

In `OptimizerBase.PrepareAndEvaluateSolution`, detect when `solution.Train()` has been called with the same inputs before and skip the redundant training. But this is a band-aid on a deeper problem.

## Why This Matters

- **Library unusable for its core use case**: training models through the facade
- **Unit tests probably pass** because they use small `MaxIterations` values or skip the optimizer
- **Production code runs 100-1000x slower than it should**
- **The 'GPU acceleration' feature is wasted** — the bottleneck is redundant training cycles, not math performance

## Reproduction

```csharp
using AiDotNet;
using AiDotNet.Data.Loaders;
using AiDotNet.Tensors.LinearAlgebra;
using AiDotNet.Optimizers;
using AiDotNet.Regression;
using AiDotNet.Models.Options;

// 2000 rows × 4 features — should solve in <1 second
var features = new Matrix<double>(2000, 4);  // filled with data
var targets = new Vector<double>(2000);       // filled with targets

var result = await new AiModelBuilder<double, Matrix<double>, Vector<double>>()
    .ConfigureDataLoader(new InMemoryDataLoader<double, Matrix<double>, Vector<double>>(features, targets))
    .ConfigureOptimizer(new AdamOptimizer<double, Matrix<double>, Vector<double>>(null,
        new AdamOptimizerOptions<double, Matrix<double>, Vector<double>>
        { InitialLearningRate = 0.01, MaxIterations = 1000 }))
    .ConfigureModel(new MultipleRegression<double>(new RegressionOptions<double> { UseIntercept = true }))
    .BuildAsync();

// Takes 5+ minutes, never completes in reasonable time
```

## Priority

This should be treated as a **critical blocker** — the library's primary training API doesn't work correctly for production use. All model types are affected. The fix requires rethinking the model/optimizer contract.

**Comment by ooples** [comment]
## Deeper Root Cause Discovered

The real issue is a mismatch between closed-form models and iterative optimizers.

`MultipleRegression.Train()` [line 109 in MultipleRegression.cs] uses the **closed-form solution**:
```csharp
public override void Train(Matrix<T> x, Vector<T> y)
{
    if (Options.UseIntercept) x = x.AddConstantColumn(NumOps.One);
    var xTx = x.Transpose().Multiply(x);
    var regularizedXTx = xTx.Add(Regularization.Regularize(xTx));
    var xTy = x.Transpose().Multiply(y);
    var solution = SolveSystem(regularizedXTx, xTy);
    // ...
}
```

This solves `(X^T X)^-1 X^T y` in ONE step — no iteration needed. The model is fully trained after one call.

But the `AdamOptimizer` wraps this in `MaxIterations=1000` loop, calling `Train()` (which does the full closed-form solve) 1000 times redundantly. Each iteration computes:
- X^T X multiplication (O(n·d²))
- Matrix inversion / linear system solve (O(d³))
- Extra work for evaluation

For 2000 rows × 4 features, that's ~32,000 multiplications per iteration × 1000 iterations = 32 million operations — and that's just the matrix multiply, plus the linear system solver.

## Suggested Fix

One of:
1. **Detect closed-form models** in `OptimizerBase` and short-circuit the loop: if the model implements `IClosedFormModel`, only call Train once
2. **Use iterative training for regression** when wrapped in an optimizer (replace closed-form with gradient descent for this path)
3. **Deprecate using optimizers with closed-form models** in `AiModelBuilder` — they shouldn't be compatible

The third option seems cleanest — if you're using `MultipleRegression`, you don't need `AdamOptimizer`.

**Comment by ooples** [comment]
## Correct Architectural Fix: Default Early-Stopping Learning Rate Scheduler

A learning rate scheduler with early stopping should be enabled by default in the optimizer. This provides the correct architectural solution:

1. For closed-form models (MultipleRegression): after epoch 1, loss won't improve → early stop triggers → training completes in ~2 iterations
2. For gradient-descent models (neural networks): continues iterating normally until loss plateaus or MaxIterations reached
3. Any optimizer can work with any model — no special-case logic needed

## Implementation

Add default early stopping to AdamOptimizer (and other optimizers):

```csharp
public class AdamOptimizerOptions<...>
{
    public int MaxIterations { get; set; } = 1000;
    public int PatienceEpochs { get; set; } = 5;  // Stop if no improvement for N epochs
    public double MinDelta { get; set; } = 1e-6;  // Minimum loss improvement to count
    public bool EnableEarlyStopping { get; set; } = true;  // DEFAULT ON
}
```

In the optimization loop:
```csharp
double bestLoss = double.MaxValue;
int epochsSinceImprovement = 0;

for (int epoch = 0; epoch < MaxIterations; epoch++)
{
    // ... existing training ...
    
    var currentLoss = currentStepData.Loss;
    if (currentLoss < bestLoss - Options.MinDelta)
    {
        bestLoss = currentLoss;
        epochsSinceImprovement = 0;
    }
    else if (Options.EnableEarlyStopping)
    {
        epochsSinceImprovement++;
        if (epochsSinceImprovement >= Options.PatienceEpochs)
        {
            // Early stop — loss hasn't improved
            break;
        }
    }
}
```

## Benefits

- MultipleRegression + AdamOptimizer stops at epoch 2-3 instead of running all 1000
- Neural networks get adaptive stopping when they've converged
- Training time becomes proportional to difficulty, not arbitrary MaxIterations value
- Same optimizer code works for closed-form and iterative models

This replaces my earlier suggestion of detecting closed-form models — early stopping is the more general solution.

**Comment by ooples** [comment]
## Critical Architectural Flaw — Library-Wide Impact

This is not just a performance bug. It's a fundamental design error that affects the entire optimizer/model relationship across the library.

## What's Happening

**AiDotNet's current architecture:**
- External optimizer (e.g., AdamOptimizer) has an `Optimize()` loop with MaxIterations epochs
- Inside each epoch: `EvaluateSolution` → `PrepareAndEvaluateSolution` → `TrainAndEvaluateSolution` → `model.Train()`
- `model.Train()` does its OWN complete training:
  - `MultipleRegression.Train()` — closed-form solve (one shot)
  - `NeuralNetworkBase.Train()` — calls `TrainWithTape` with internal tape-based backprop
- So every "epoch" of the external optimizer triggers a complete from-scratch training

**Result:** The model trains itself 1000 times when MaxIterations=1000, each producing the same result.

## How PyTorch Does It (Correctly)

In PyTorch:
```python
model = nn.Linear(4, 1)           # Model is just a FUNCTION, no Train() method
optimizer = torch.optim.Adam(model.parameters(), lr=0.01)

for epoch in range(num_epochs):
    predictions = model(x)        # Forward pass
    loss = loss_fn(predictions, y) # Compute loss
    loss.backward()               # Compute gradients
    optimizer.step()              # Update parameters
    optimizer.zero_grad()         # Reset gradients
```

Key insights:
1. The model has NO `train()` method that does optimization
2. The optimizer ONLY updates parameters — it doesn't know how to train
3. The user writes the loop
4. For closed-form regression, you use `torch.linalg.lstsq(x, y)` directly — no optimizer involved

## The Root Fix

Models should NOT have `Train()` methods that do their own optimization. Instead:

1. **Models** should expose:
   - `Forward(x)` → predictions
   - `Backward(gradient)` → parameter gradients
   - `GetParameters()` / `SetParameters()`

2. **Optimizers** should:
   - Call `model.Forward(x)` to get predictions
   - Compute loss
   - Call `model.Backward(loss_gradient)` to get parameter gradients  
   - Apply the update (Adam, SGD, etc.) to parameters

3. **Closed-form models** (MultipleRegression, RidgeRegression):
   - Have their own `Solve(x, y)` method (not Train)
   - Bypass optimizers entirely — the builder should detect this and call `Solve` directly

## Affected Components

Every combination of (external optimizer) × (model with internal Train) is affected:
- `AiModelBuilder` → `AdamOptimizer` → `MultipleRegression` (closed form gets redone 1000x)
- `AiModelBuilder` → `AdamOptimizer` → `Transformer` (NN trains itself internally, external Adam is wasted work)
- `AiModelBuilder` → any optimizer → any model with internal `Train()`

## Evidence

- Week 1 AI training course homework: a basic regression on 2000 rows takes 5+ minutes with MaxIterations=1000 (should be <1 second)
- Same model with MaxIterations=50 still doesn't complete in reasonable time
- The "GPU acceleration" the library advertises doesn't help because the bottleneck is the redundant training cycles, not the math

## Recommended Path Forward

1. Add an `IClosedFormModel` interface marker
2. `AiModelBuilder.BuildAsync()` checks if model implements it — if so, call `model.Solve()` directly, skip optimizer entirely
3. For iterative models, redesign so `Optimize()` does NOT call `model.Train()` inside its loop — instead, it should only do Forward/Backward/Update using the model as a function
4. The existing `TrainWithTape` pattern is correct — it should be what runs for NNs, with the external optimizer loop REMOVED

This is a breaking change but it's necessary — the current architecture can't produce correct training results efficiently for any model type.
